api/app/signals/apps/signals/v1/views.py

```diff
@@ -54,13 +57,41 @@ class PrivateSignalViewSet(DatapuntViewSet):
     authentication_classes = (JWTAuthBackend, )
     filter_backends = (DjangoFilterBackend, )
 
-    @action(detail=True)  # default GET gets routed here
+    @action(detail=True)
     def history(self, request, pk=None):
+        """History endpoint filterable by action."""
         history_entries = History.objects.filter(_signal__id=pk)
-        serializer = HistoryHalSerializer(history_entries, many=True)
+        what = self.request.query_params.get('what', None)
+        if what:
+            history_entries = history_entries.filter(what=what)
 
+        serializer = HistoryHalSerializer(history_entries, many=True)
         return Response(serializer.data)
 
+    # https://stackoverflow.com/questions/45564130/django-rest-framework-image-upload
+    # note starting DRF 3.8, @detail_route and @list_route are replaced with action
+
+    @action(detail=True, methods=['POST'])
+    def image(self, request, pk=None):
+        signal = Signal.objects.get(pk=pk)
+
+        if signal.image:
+            raise PermissionDenied("Melding is reeds van foto voorzien.")
```

Comment:
PermissionDenied does not seem to fit the exception maybe MethodNotAllowed is a better fit because the method is not allowed when there is already an image attached to the signal

Reply:
agree